### ISSUE TITLE
Predicton step that can be run in between training epochs

### ISSUE THREAD

**Issue by OlfwayAdbayIgbay** [PyTorchLightning/pytorch-lightning#8808]
## 🚀 Feature
Add an "inference" step, or a keyword to trainer.predict() which doesn't remove the model from GPU after the prediction run.

### Motivation

For things like online triplet mining, I am predicting batches of data in between epochs. I don't need gradients for those, so I use self.trainer.predict.

something curious I've noted, is that after running "trainer.predict", the model is removed from the cuda device. I assume that is wanted behaviour, so right now I fix that by running "self.cuda(gpu_index)" afterwards.

### Pitch

Since lightning is all about keeping code beautiful and organized, I propose there should be a keyword in trainer.predict() which keeps the model on the cuda device.

### Alternatives

Right now, it works to write "self.cuda(gpu_index)"  at the end of the prediction loop. However, if one doesn't know this, then it can take quite a while to figure out, especially in larger projects.

**Comment by tchaton** [comment]
Dear @OlfwayAdbayIgbay,

This sounds reasonable. Potentially we could prevent this behaviour for predict only.

@ananthsub Any opinions ?

Best,
T.C

**Comment by ananthsub** [comment]
@OlfwayAdbayIgbay could you share what your code calling trainer.predict currently looks like? I'm curious about this

> For things like online triplet mining, I am predicting batches of data in between epochs. I don't need gradients for those, so I use self.trainer.predict.


@tchaton i don't think this is prediction specific, right? this is more about the accelerator/training type, and expectations around moving the model back to cpu after execution. this is currently hardcoded in the parallel plugin: https://github.com/PyTorchLightning/pytorch-lightning/blob/c4a1c8ba20975ecf5d3ffb76b8cbc107ab761356/pytorch_lightning/plugins/training_type/parallel.py#L141-L145

In general I wonder what our principles for plugins are in terms of cleanup. We had a slightly related discussion with @awaelchli, @carmocca, and @yifuwang about process groups here: https://github.com/PyTorchLightning/pytorch-lightning/pull/8080

Strawman proposal: if the model is already on the device at the start of training, we don't move it to CPU afterwards. This way, the lightning trainer leaves things as it finds them. is that too idealistic?

**Comment by awaelchli** [comment]
> Strawman proposal: if the model is already on the device at the start of training, we don't move it to CPU afterwards. This way, the lightning trainer leaves things as it finds them. is that too idealistic?

makes sense to me :) just requires more bookkeeping for metrics, optimizer state, and model parameters (and maybe more) too keep track of.

**Comment by OlfwayAdbayIgbay** [comment]
@ananthsub of course! See the following:

It also has to be done on_train_epoch_start, because if you call this on_train_epoch_end, the trainer will not do the logging! Maybe I should open an issue for that as well :) 

```
    def on_train_epoch_start(self):
        # Gets random triplets
        self.dataset.start_building()
        # Predict Embeddings
        embeddings = self.trainer.predict(self,return_predictions=True)
        embeddings = torch.vstack(embeddings)
        # Sort by class
        embeddings = embeddings.view(self.dataset.num_classes,len(embeddings)//self.dataset.num_classes,embeddings.shape[-1])

        #Mine the embeddings (done on cpu, for maximum gpu 
        self.dataset.start_mining(embeddings.cpu())

        #Put model back on gpu, which has been moved off the gpu after trainer.predict ended
        self.cuda(*gpu_index)


```
